### PR TITLE
Improve GDScript autocompletion for methods

### PR DESCRIPTION
**Pull request by Lazy-Rabbit-2001**
Implements the part 1 of https://github.com/godotengine/godot-proposals/issues/11079
Check the link for details about the reason why this pr is posted.

### CODE REVIEW

**Comment by Calinou** [comment]
> Kinda off-topic, but it surprised me that the back braces will not get overlapped.

It depends on whether you accept the suggestion with <kbd>Tab</kbd> or <kbd>Enter</kbd>, like in many other editors. See https://github.com/godotengine/godot/pull/90723 which changes this behavior to be more predictable.

File: modules/gdscript/gdscript_editor.cpp [hunk 1]
```diff
@@ -1115,10 +1115,12 @@ static void _find_identifiers_in_class(const GDScriptParser::ClassNode *p_class,
 							continue;
 						}
 						option = ScriptLanguage::CodeCompletionOption(member.function->identifier->name, ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION, location);
-						if (member.function->parameters.size() > 0) {
+						if (member.function->parameters.size() > 0 || bool((member.function->info.flags & METHOD_FLAG_VARARG) >> 4)) {
```
Comment (HolonProduction) [inline modules/gdscript/gdscript_editor.cpp:1118]: What's the deal with this right shift? Wouldn't `(member.function->info.flags & METHOD_FLAG_VARARG)` do the same?

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Code looks good to me.

image

This is a nice usability improvement 🙂

**Comment by AdriaandeJongh** [comment]
Considering the autocomplete popup uses a monospaced font, and the three dots inside function parentheses take up a considerable width, may I suggest using this character instead? 

`…`

**Comment by HolonProduction** [comment]
This is indeed a nice improvement. I think the usage of `…` works better:

image

Looks good to me on code and functionality side. You still need to [squash the commits](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) before this can get merged.

**Comment by HolonProduction** [comment]
Ok, now it's squashed, but could we get a meaningful commit name? E.g. just the title of the PR. If you already squashed you can use amend for it:
```bash
git commit --amend -m "Improve GDScript autocompletion for methods"
git push -f
```

File: modules/gdscript/gdscript_editor.cpp [hunk 2]
```diff
@@ -1139,10 +1139,12 @@ static void _find_identifiers_in_class(const GDScriptParser::ClassNode *p_class,
 							continue;
 						}
 						option = ScriptLanguage::CodeCompletionOption(member.function->identifier->name, ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION, location);
-						if (member.function->parameters.size() > 0) {
+						if (member.function->parameters.size() > 0 || member.function->info.flags & METHOD_FLAG_VARARG) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1142]:
```suggestion
						if (member.function->parameters.size() > 0 || (member.function->info.flags & METHOD_FLAG_VARARG)) {
```
For readability

File: modules/gdscript/gdscript_editor.cpp [hunk 3]
```diff
@@ -1241,10 +1244,12 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 							}
 							int location = p_recursion_depth + _get_method_location(scr->get_class_name(), E.name);
 							ScriptLanguage::CodeCompletionOption option(E.name, ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION, location);
-							if (E.arguments.size()) {
+							if (E.arguments.size() || E.flags & METHOD_FLAG_VARARG) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1247]:
```suggestion
							if (E.arguments.size() || (E.flags & METHOD_FLAG_VARARG)) {
```

File: modules/gdscript/gdscript_editor.cpp [hunk 4]
```diff
@@ -1327,10 +1332,12 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 					}
 					int location = p_recursion_depth + _get_method_location(type, E.name);
 					ScriptLanguage::CodeCompletionOption option(E.name, ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION, location);
-					if (E.arguments.size()) {
+					if (E.arguments.size() || E.flags & METHOD_FLAG_VARARG) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1335]:
```suggestion
					if (E.arguments.size() || (E.flags & METHOD_FLAG_VARARG)) {
```

File: modules/gdscript/gdscript_editor.cpp [hunk 5]
```diff
@@ -1398,10 +1405,12 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 						continue;
 					}
 					ScriptLanguage::CodeCompletionOption option(E.name, ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION, location);
-					if (E.arguments.size()) {
+					if (E.arguments.size() || E.flags & METHOD_FLAG_VARARG) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1408]:
```suggestion
					if (E.arguments.size() || (E.flags & METHOD_FLAG_VARARG)) {
```

File: modules/gdscript/gdscript_editor.cpp [hunk 6]
```diff
@@ -1431,10 +1440,12 @@ static void _find_identifiers(const GDScriptParser::CompletionContext &p_context
 	for (const StringName &E : functions) {
 		MethodInfo function = GDScriptUtilityFunctions::get_function_info(E);
 		ScriptLanguage::CodeCompletionOption option(String(E), ScriptLanguage::CODE_COMPLETION_KIND_FUNCTION);
-		if (function.arguments.size() || (function.flags & METHOD_FLAG_VARARG)) {
+		if (function.arguments.size() || function.flags & METHOD_FLAG_VARARG) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1443]:
```suggestion
		if (function.arguments.size() || (function.flags & METHOD_FLAG_VARARG)) {
```

**Review by AThousandShips — APPROVED** [review]

**Review by HolonProduction — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first merged contribution! 🎉

**Comment by Meorge** [comment]
I'm sorry to be a bit of a downer on this, but at least would there be a way to change this behavior depending on whether or not Godot's internal editor is receiving the suggestions versus an external IDE? The new behavior is inconsistent with other languages in VS Code, where the parentheses are never given after brackets in the suggestion list or the actual completed code.

Currently, for GDScript, starting to type out `print`:
CleanShot 2024-11-25 at 22 16 04@2x
Immediately after hitting Enter/Return:
CleanShot 2024-11-25 at 22 16 19@2x

By contrast, here are some other popular languages in VS Code:

For Python, starting to type out `print`:
CleanShot 2024-11-25 at 22 08 19@2x
Immediately after hitting Enter/Return:
CleanShot 2024-11-25 at 22 08 41@2x

For TypeScript, starting to type out `console.log`:
CleanShot 2024-11-25 at 22 09 35@2x
Immediately after hitting Enter/Return:
CleanShot 2024-11-25 at 22 09 43@2x

For C/C++, starting to type out `print_line`:
CleanShot 2024-11-25 at 22 13 33@2x
Immediately after hitting Enter/Return:
CleanShot 2024-11-25 at 22 13 43@2x

It seems the standard is to not include parentheses in the suggestions for function/method names, and to not insert them once the user accepts the suggestion. Perhaps it makes more sense for Godot's built-in editor, but for external ones such as VS Code it looks to be different. IMO it'd make sense to either change the autocomplete behavior depending on the recipient (built-in editor vs external editor), or change the built-in editor's behavior so that it is more consistent with other language servers.

**Comment by HolonProduction** [comment]
I think only referencing language servers developed by Microsoft might be a bit biased (I suppose you are using the Microsoft c++ language server).

Here are some counter examples:

clangd C++ language server:
Screenshot From 2024-11-26 11-07-13
Screenshot From 2024-11-26 11-07-23

Dart integration
image
image

Java (the oracle plugin)
image
image

**Comment by Meorge** [comment]
Thanks for providing the counterexamples! You're right that they were all Microsoft-developed servers, as far as I know - my thought was that since Microsoft originally developed the specification, if there was a standard for how/when to do things like insert parentheses, the Microsoft servers would be the most likely to follow it to a T.

The [documentation for VS Code's Rust support on their website](https://code.visualstudio.com/docs/languages/rust) features `rust-analyzer`, which also shows `(...)` after function/method names 🙂 



Given that these other big language servers do it as well, I agree having the GDScript server do it is all good. 😄

**Pull request merged**